torch/lib/THCUNN/generic/FusedRNNKernel.cu

```diff
@@ -0,0 +1,963 @@
+    gradoutcI.collapseDims();
+    gradinI.collapseDims();
+
+    FILL_DIM(unsigned long, maxDim, LSTM_BACKWARD);
+
+  }
+  THCudaCheck(cudaGetLastError());
+}
+
+void THNN_(GRUFused_updateOutput)(
+          THCState *state,
+          THCTensor *input,
+	  THCTensor *hidden,
+	  THCTensor *bias1,
+	  THCTensor *bias2,
+	  THCTensor *hx,
+	  THCTensor *hy)
+{
+  THCTensor_(resizeAs)(state, hy, hx);
+  THNN_(FusedRNNAssertSizes)(state, 3, 4, input, hidden, hx, hy);
+
+  bool has_bias = (bias1!=NULL);
+  bool canUse32bi;
+  int maxDim;
+
+  if(has_bias){
+
+    THCUNN_assertSameGPU
+      (state, 6, input, hidden, hx, hy, bias1, bias2);
+    canUse32bi = THNN_(canUse32BitIndexMath)
+      (state, 6, input, hidden, hx, hy, bias1, bias2);
+    maxDim = THNN_(minIndexType)
+      (state, 6, input, hidden, hx, hy, bias1, bias2);
+
+    TensorInfo<DATATYPE, unsigned long> tmphi =
+      getTensorInfo<THCTensor, unsigned long>(state, hx);
+    unsigned long tmp_hid = tmphi.sizes[tmphi.dims-1];
+    
+    THAssertMsg( tmp_hid*3 == THCTensor_(nElement)(state, bias1) &&
+		 tmp_hid*3 == THCTensor_(nElement)(state, bias2),
+		 "Bias in pointwise operation is an incorrect size, must be 3 x feature size.");
+  }else{
+    THCUNN_assertSameGPU
+      (state, 4, input, hidden, hx, hy);
+    canUse32bi = THNN_(canUse32BitIndexMath)
+      (state, 4, input, hidden, hx, hy);
+    maxDim = THNN_(minIndexType)
+      (state, 4, input, hidden, hx, hy);
+  }
+
+  const dim3 block = getApplyBlock();
+  //const dim3 block(32, 32);
+  ptrdiff_t totalElements = TensorUtils<THCTensor>::getNumElements(state, hx);
+
+  dim3 grid;
+
+  THAssertMsg(getApplyGrid(state, totalElements, grid),
+	      "Could not get grid size for pointwise apply");
+  if(canUse32bi){
+    TensorInfo<DATATYPE, unsigned int> inputI =
+      getTensorInfo<THCTensor, unsigned int>(state, input);
+    TensorInfo<DATATYPE, unsigned int> hiddenI =
+      getTensorInfo<THCTensor, unsigned int>(state, hidden);
+    TensorInfo<DATATYPE, unsigned int> hxI =
+      getTensorInfo<THCTensor, unsigned int>(state, hx);
+    TensorInfo<DATATYPE, unsigned int> hyI =
+      getTensorInfo<THCTensor, unsigned int>(state, hy);
+
+    unsigned int hid_size = hxI.sizes[hxI.dims-1];
+
+    inputI.collapseDims();
+    hiddenI.collapseDims();
+    hyI.collapseDims();
+    hxI.collapseDims();
+
+    unsigned int zero[1] = {0};
+    TensorInfo<DATATYPE, unsigned int> nullinfo =
+      TensorInfo<DATATYPE, unsigned int>(NULL, 1, zero, zero);
+    TensorInfo<DATATYPE, unsigned int> bias1I = nullinfo;
+    TensorInfo<DATATYPE, unsigned int> bias2I = nullinfo;
+
+    if(has_bias){
+      bias1I = getTensorInfo<THCTensor, unsigned int>(state, bias1);
+      bias2I = getTensorInfo<THCTensor, unsigned int>(state, bias2);
+      bias1I.collapseDims();
+      bias2I.collapseDims();
+    }
+
+    FILL_DIM(unsigned int, maxDim, GRU_FORWARD);
+      
+  }else{
+
+    TensorInfo<DATATYPE, unsigned long> inputI =
```

Comment:
There's a lot of duplication in these cases (in all top level functions), and it seems that the only difference is `unsigned long` vs `unsigned int`. Why not separate them into a templated functions and instantiate different versions in each case?